pkg/storage/stores/indexshipper/compactor/deletion/request_handler.go

```diff
@@ -114,25 +222,25 @@ func (dm *DeleteRequestHandler) CancelDeleteRequestHandler(w http.ResponseWriter
 
 	params := r.URL.Query()
 	requestID := params.Get("request_id")
-
-	deleteRequest, err := dm.deleteRequestsStore.GetDeleteRequest(ctx, userID, requestID)
+	deleteRequests, err := dm.deleteRequestsStore.GetDeleteRequestGroup(ctx, userID, requestID)
 	if err != nil {
+		if errors.Is(err, ErrDeleteRequestNotFound) {
+			http.Error(w, "could not find delete request with given id", http.StatusNotFound)
+			return
+		}
+
 		level.Error(util_log.Logger).Log("msg", "error getting delete request from the store", "err", err)
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	if deleteRequest == nil {
-		http.Error(w, "could not find delete request with given id", http.StatusBadRequest)
-		return
-	}
-
-	if deleteRequest.Status != StatusReceived {
+	toDelete := filterProcessed(deleteRequests)
+	if len(toDelete) == 0 {
 		http.Error(w, "deletion of request which is in process or already processed is not allowed", http.StatusBadRequest)
 		return
 	}
 
-	if err := dm.deleteRequestsStore.RemoveDeleteRequest(ctx, userID, requestID, deleteRequest.CreatedAt, deleteRequest.StartTime, deleteRequest.EndTime); err != nil {
+	if err := dm.deleteRequestsStore.RemoveDeleteRequests(ctx, toDelete); err != nil {
```

Comment:
Because of the requirement to recover from a delete request that cannot complete, I think we need the ability to cancel requests at any time. I've added `% Complete` to requests that are partially processed to provide some more visibility to the user. The result is:

```
[
  {
    "request_id": "626dd400",
    "start_time": 1660653094,
    "end_time": 1660663894,
    "query": "{job=\"loki-ops/ingester\"} |= \"warn\"",
    "status": "66% Complete",
    "created_at": 1660770050.104
  }
]
```

What do you think?

Reply:
Yeah, showing `%` in status looks like a good idea. The problem with partial cancellation, though, is that there is a chance that it already would have marked chunks from some of the tables for deletion, so it won't be a clean cancellation. Also, since the compactor already picked up the request for processing, you would have to increase the cancellation period config and restart the compactor.

So I wonder if we should add a query param to the cancellation request to explicitly ask the compactor to force cancel the request and document that it won't be a clean cancellation and some data might already have been marked for deletion. However, it would still require the compactor to restart to not continue processing the delete request since we can't interrupt the ongoing compaction, which I think should be okay for now. Other than documentation, it would also save us from doing the config change step.

What do you think?

Reply:
:+1: I think that works. The latest commit disallows cancellation of partially-completed deletes unless a `force=true` query parameter is passed.